src/OAuthClient.php

```diff
@@ -0,0 +1,157 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ * ---------------------------------------------------------------------
+ */
+
+use Glpi\Application\View\TemplateRenderer;
+use Glpi\OAuth\Server;
+
+final class OAuthClient extends CommonDBTM
+{
+    public static $rightname = 'oauth_client';
+
+    public static $undisclosedFields = [
+        'secret',
+    ];
+
+    public static function getTypeName($nb = 0)
+    {
+        return _n('OAuth client', 'OAuth clients', $nb);
+    }
+
+    public static function getIcon()
+    {
+        return 'ti ti-key';
+    }
+
+    public function showForm($ID, array $options = [])
+    {
+        TemplateRenderer::getInstance()->display('pages/setup/oauthclient.html.twig', [
+            'item' => $this,
+            'params' => $options,
+            'allowed_scopes' => Server::getAllowedScopes(),
+        ]);
+        return true;
+    }
+
+    public function rawSearchOptions()
+    {
+        $opts = [];
+
+        $opts[] = [
+            'id' => 'common',
+            'name' => self::getTypeName(1),
+        ];
+        $opts[] = [
+            'id' => 1,
+            'table' => self::getTable(),
+            'field' => 'name',
+            'name' => __('Name'),
+            'datatype' => 'itemlink',
+        ];
+        $opts[] = [
+            'id'                 => '2',
+            'table'              => $this->getTable(),
+            'field'              => 'id',
+            'name'               => __('ID'),
+            'massiveaction'      => false, // implicit field is id
+            'datatype'           => 'number'
+        ];
+
+        $opts[] = [
+            'id' => '3',
+            'table' => self::getTable(),
+            'field' => 'identifier',
+            'name' => __('Client ID'),
+            'datatype' => 'itemlink',
+        ];
+
+        return $opts;
+    }
+
+    /**
+     * @throws Exception
+     */
+    public static function getNewIDOrSecret()
+    {
+        return bin2hex(random_bytes(Server::ID_SECRET_LENGTH_BYTES));
+    }
+
+    public function prepareInputForAdd($input)
+    {
+        $key = new GLPIKey();
+        $input['identifier'] = self::getNewIDOrSecret();
+        $input['secret'] = $key->encrypt(self::getNewIDOrSecret());
+        if (!isset($input['grants'])) {
+            $input['grants'] = exportArrayToDB([]);
+        }
+        if (!isset($input['scopes'])) {
+            $input['scopes'] = exportArrayToDB([]);
+        }
+        if (!isset($input['redirect_uri'])) {
+            $input['redirect_uri'] = '/api.php/oauth2/redirection';
+        }
+
+        return $input;
+    }
+
+    public function prepareInputForUpdate($input)
+    {
+        $key = new GLPIKey();
+        if (isset($input['secret'])) {
+            $input['secret'] = $key->encrypt($input['secret']);
```

Comment:
Is it a good idea to make it decryptable? Most OAuth provider are only displaying this code once, and I guess they are hashing them.

If we want to make it decryptable, it should be added to `GLPIKey::$fields`, to make sure it is updated when encryption key is regenerated.

Reply:
We discussed about this in our weekly meeting and we concluded that we could keep the current logic based on decryptable secrets.

Only required change here is to register this field in `GLPIKey::$fields` to ensure value will be correctly re-encrypted when encryption key is changed.